pydantic_ai_slim/pydantic_ai/models/bedrock.py

```diff
@@ -367,39 +367,55 @@ def _map_inference_config(
     async def _map_messages(
         self, messages: list[ModelMessage]
     ) -> tuple[list[SystemContentBlockTypeDef], list[MessageUnionTypeDef]]:
-        """Just maps a `pydantic_ai.Message` to the Bedrock `MessageUnionTypeDef`."""
+        """Maps a `pydantic_ai.Message` to the Bedrock `MessageUnionTypeDef`.
+
+        Groups consecutive ToolReturnPart objects into a single user message as required by Bedrock Claude/Nova models.
+        """
         system_prompt: list[SystemContentBlockTypeDef] = []
         bedrock_messages: list[MessageUnionTypeDef] = []
         document_count: Iterator[int] = count(1)
         for m in messages:
             if isinstance(m, ModelRequest):
-                for part in m.parts:
+                parts = m.parts
+                i = 0
+                while i < len(parts):
+                    part = parts[i]
                     if isinstance(part, SystemPromptPart):
                         system_prompt.append({'text': part.content})
+                        i += 1
                     elif isinstance(part, UserPromptPart):
                         bedrock_messages.extend(await self._map_user_prompt(part, document_count))
+                        i += 1
                     elif isinstance(part, ToolReturnPart):
-                        assert part.tool_call_id is not None
-                        bedrock_messages.append(
-                            {
-                                'role': 'user',
-                                'content': [
-                                    {
-                                        'toolResult': {
-                                            'toolUseId': part.tool_call_id,
-                                            'content': [{'text': part.model_response_str()}],
-                                            'status': 'success',
-                                        }
+                        # Group consecutive ToolReturnParts
+                        tool_results: list[dict[str, typing.Any]] = []
+                        while i < len(parts) and isinstance(parts[i], ToolReturnPart):
```

Comment:
To keep this logic easier to follow, can you please move the joining of `user` parts to a post-processing step like we do in Mistral for a different issue? I assume we'll want this for any back-to-back user message, not just `toolResults`?

https://github.com/pydantic/pydantic-ai/blob/7d613c5786aaad830970aec1af7006d4cc747211/pydantic_ai_slim/pydantic_ai/models/mistral.py#L489-L499